crates/project-model/src/project_json.rs

```diff
@@ -87,6 +87,21 @@ pub struct Crate {
     pub(crate) exclude: Vec<AbsPathBuf>,
     pub(crate) is_proc_macro: bool,
     pub(crate) repository: Option<String>,
+    pub build_info: Option<BuildInfo>,
```

Comment:
@davidbarsky If I'm reading this right, you want the runnables to be stored at the root of the project.json. That will require:

- a substitution mechanism like `$label` so that the "runnable templates" can be turned into real commands for a specific crate
- probably RA using the `build_info.kind` field to figure out which runnables to offer for a given target

I also think we also want custom runnables in `build_info` for specific crates anyway. Right? Because the build system might simply know better. Substitution is nice because it saves a few MB of JSON. But ultimately you want the ability to add more to specific crates.

Because I believe you're going to want runnables per-crate anyway, and they are completely capable of handling anything substitution can, I think we can merge this PR pretty much as-is. I think the substitution mechanism is more of an optimisation than anything else.

I am happy to put together a PR that allows flycheck to use these.

Reply:
@cormacrelf 

> a substitution mechanism like `$label` so that the "runnable templates" can be turned into real commands for a specific crate

Yup: the substitution mechanism is thankfully not _too_ complex: I think it boils down to *just* a [`str::replace`](https://doc.rust-lang.org/stable/std/string/struct.String.html#method.replace).

(If you squint, I'd argue that the substitution already exists in rust-analyzer today via Cargo, but the substitution knowledge is pretty implicit...)

> probably RA using the `build_info.kind` field to figure out which runnables to offer for a given target

Will need to repage some context in, but I'm pretty sure the answer is also "yup".

> I also think we also want custom runnables in `build_info` for specific crates anyway. Right? Because the build system might simply know better. Substitution is nice because it saves a few MB of JSON. But ultimately you want the ability to add more to specific crates.

I don't know if I agree: while I'm assuming that while there *is* a smart build system, I don't think there isn't much surface area for more specialized commands for individual crates. There's also the fact that rust-analyzer is pretty darn fast even compared to buck2, which is the fastest build system I'm aware of—single digit milliseconds versus ~300 milliseconds. There's also the matter of the duplication being somewhat concerning for the [Fuschia folks](https://github.com/rust-lang/rust-analyzer/issues/15892#issuecomment-1919859373), as their `rust-project.json` is already pretty large and I'm not really inlined to make things more complicated for them, however slightly.

> Because I believe you're going to want runnables per-crate anyway, and they are completely capable of handling anything substitution can, I think we can merge this PR pretty much as-is. I think the substitution mechanism is more of an optimisation than anything else.

You are right it's an optimization, but it's a pretty cheap one. Wilfred and I need to get this over the finish line, but it's genuinely not too much work to implement it.

Reply:
OK. A good metric for the design is whether `rust-project.json` can encode all the cargo-related logic. If you set `build_info.label` to a cargo crate name, and a root level fly check template of `cargo check -p $label --message-format=json`, then we almost have that mission in the bag. The main missing piece is flycheck for a crate vs for the whole workspace, which is configurable through `check.workspace` in RA. Having two more runnable kinds, `flycheck` and `flycheckWorkspace`, should do it.